docs/standard/io/file-path-formats.md

```diff
@@ -0,0 +1,171 @@
+---
+title: "File path formats on Windows systems"
+ms.date: "04/25/2018"
+ms.prod: ".net"
+ms.technology: dotnet-standard
+ms.topic: "article"
+helpviewer_keywords: 
+  - "I/O, long paths"
+  - "long paths"
+  - "path formats, Windows"
+author: "rpetrusha"
+ms.author: "ronpet"
+ms.workload: 
+  - "dotnet"
+  - "dotnetcore"
+---
+# File path formats on Windows systems
+
+Members of many of the types in the <xref:System.IO> namespace include a `path` parameter that lets you specify an absolute or relative path to a file system resource. This path is then passed to [Windows file sysem APIs](https://msdn.microsoft.com/library/windows/desktop/aa364407(v=vs.85).aspx). This topic discusses the formats for file paths that you can use on Windows systems.
+
+## Traditional DOS paths
+
+A standard DOS path can consist of three components:
+
+- A volume or drive letter followed by the volume separator (`:`).
+- A directory name. The [directory separator character](<xref:System.IO.Path.DirectorySeparatorChar>) separates subdirectories within the nested directory hierarchy.
+- An optional filename. The [directory separator character](<xref:System.IO.Path.DirectorySeparatorChar>) separates the file path and the filename.
+
+If all three components are present, the path is absolute. If no volume or drive letter is specified and the directory names begins with the [directory separator character](<xref:System.IO.Path.DirectorySeparatorChar>), the path is relative from the root of the current drive is assumed. Otherwise, the path is relative to the current directory. The following table shows some possible directory and file paths.
```

Comment:
"the path is relative from the root of the current drive is assumed" reads a strange to me.

Reply:
The super important case is `C:foo` as opposed to `C:\foo`. You describe it below, I think it might be worth mentioning here too. It is the biggest gotcha in Windows paths. One possible way to explain it is to separate the path into two pieces:

1. The (optional) volume specifier (drive letter A-Z + colon)
2. The directory/file portion

If the directory/file portion begins with a directory separator, it is relative to the root of either the specified drive, or, if the drive is not specified, the drive of the current working directory `Environment.CurrentDirectory` or `Directory.GetCurrentDirectory()`. 

If the directory/file portion doesn't begin with a separator, it is relative to the last known current directory set for specified drive, or, if the drive is not specified, the full path of the current working directory `Environment.CurrentDirectory` or `Directory.GetCurrentDirectory()`. 

The term we're using in the API is "fully qualified". `Path.IsPathFullyQualified()` will tell you if the path is independent of the current directory (i.e. will never change if the current directory changes). The path may contain relative directory segments (`.` or `..`) and be fully qualified, as the resolved path will always point to the same location.

Here some sample code showing the drive-relative resolution:

``` C#
// Example presumes E:\Bar exists.
// You can't change the current directory to paths that don't exist.

Directory.SetCurrentDirectory(@"C:\");
string path = Path.GetFullPath(@"E:\Foo");
Console.WriteLine($"'E:\\Foo' resolves to {path}");
path = Path.GetFullPath(@"E:Foo");
Console.WriteLine($"'E:Foo' resolves to {path}");

Console.WriteLine("Setting current directory to 'E:\\Bar'");
Directory.SetCurrentDirectory(@"E:\Bar");

path = Path.GetFullPath(@"E:\Foo");
Console.WriteLine($"'E:\\Foo' resolves to {path}");
path = Path.GetFullPath(@"E:Foo");
Console.WriteLine($"'E:Foo' resolves to {path}");
```

Output:

```
'E:\Foo' resolves to E:\Foo
'E:Foo' resolves to E:\Foo
Setting current directory to 'E:\Bar'
'E:\Foo' resolves to E:\Foo
'E:Foo' resolves to E:\Bar\Foo
```

Reply:
Ok, I was missing one part in the example. Drive relative only considers the path for the actual current directory or last current directory set by the cmd prompt. Only the cmd prompt sets the hidden environment variables that store the last set current directory.

Here is the more complicated, but complete example:

``` C#
// Example presumes E:\Bar exists and you haven't set any current directory
// from the command prompt for E:\ before launching.
//
// (You can't change the current directory to paths that don't exist.)

Console.WriteLine($"Current directory is '{Environment.CurrentDirectory}'");
Console.WriteLine("Setting current directory to 'C:\\'");
Directory.SetCurrentDirectory(@"C:\");
string path = Path.GetFullPath(@"E:\Foo");
Console.WriteLine($"'E:\\Foo' resolves to {path}");
path = Path.GetFullPath(@"E:Foo");
Console.WriteLine($"'E:Foo' resolves to {path}");

Console.WriteLine("Setting current directory to 'E:\\Bar'");
Directory.SetCurrentDirectory(@"E:\Bar");

path = Path.GetFullPath(@"E:\Foo");
Console.WriteLine($"'E:\\Foo' resolves to {path}");
path = Path.GetFullPath(@"E:Foo");

// This will be "E:\Bar\Foo" as it happens to match the drive of the current directory
Console.WriteLine($"'E:Foo' resolves to {path}");

Console.WriteLine("Setting current directory to 'C:\\'");
Directory.SetCurrentDirectory(@"C:\");

path = Path.GetFullPath(@"E:\Foo");
Console.WriteLine($"'E:\\Foo' resolves to {path}");

// This will be either "E:\Foo" or "E:\Foo\Foo" in the subprocess. In the sub process,
// the command prompt set the current directory before launch of our application, which
// sets a hidden environment variable that is considered.
path = Path.GetFullPath(@"E:Foo");
Console.WriteLine($"'E:Foo' resolves to {path}");

if (args.Length < 1)
{
    Console.WriteLine(@"Launching again, after setting to E:\Foo");
    Uri currentExe = new Uri(Assembly.GetExecutingAssembly().GetName().CodeBase, UriKind.Absolute);
    string commandLine = $"/C cd E:\\Foo & \"{currentExe.LocalPath}\" stop";
    ProcessStartInfo psi = new ProcessStartInfo("cmd", commandLine); ;
    Process.Start(psi).WaitForExit();

    Console.WriteLine("Sub process returned:");
    path = Path.GetFullPath(@"E:\Foo");
    Console.WriteLine($"'E:\\Foo' resolves to {path}");
    path = Path.GetFullPath(@"E:Foo");
    Console.WriteLine($"'E:Foo' resolves to {path}");
}

Console.Write("Press enter to continue:");
Console.ReadLine();
```

First output:

```
Current directory is 'C:\Users\jerem\Documents\Visual Studio 2017\Projects\ConsoleApp17\ConsoleApp17\bin\Debug'
Setting current directory to 'C:\'
'E:\Foo' resolves to E:\Foo
'E:Foo' resolves to E:\Foo
Setting current directory to 'E:\Bar'
'E:\Foo' resolves to E:\Foo
'E:Foo' resolves to E:\Bar\Foo
Setting current directory to 'C:\'
'E:\Foo' resolves to E:\Foo
'E:Foo' resolves to E:\Foo
Launching again, after setting to E:\Foo
Sub process returned:
'E:\Foo' resolves to E:\Foo
'E:Foo' resolves to E:\Foo
```

Sub launched process:

```
Current directory is 'C:\'
Setting current directory to 'C:\'
'E:\Foo' resolves to E:\Foo
'E:Foo' resolves to E:\Foo\Foo
Setting current directory to 'E:\Bar'
'E:\Foo' resolves to E:\Foo
'E:Foo' resolves to E:\Bar\Foo
Setting current directory to 'C:\'
'E:\Foo' resolves to E:\Foo
'E:Foo' resolves to E:\Foo\Foo
```

Another subtlety here is that environment variables are copied from the spawning process. Hence when we check the state again after returning we don't get 'E:\Foo\Foo`.